SCons/Tool/ninja.py

```diff
@@ -1462,68 +1483,43 @@ def robust_rule_mapping(var, rule, tool):
+    ninja_syntax = importlib.import_module(".ninja_syntax", package='ninja')
         
     if NINJA_STATE is None:
         NINJA_STATE = NinjaState(env, ninja_file[0], ninja_syntax.Writer)
-        NINJA_STATE.ninja_bin_path = os.path.abspath(os.path.join(ninja_syntax.__file__, os.pardir, 'data', 'bin'))
-    # Here we will force every builder to use an emitter which makes the ninja
-    # file depend on it's target. This forces the ninja file to the bottom of
-    # the DAG which is required so that we walk every target, and therefore add
-    # it to the global NINJA_STATE, before we try to write the ninja file.
-    def ninja_file_depends_on_all(target, source, env):
-        if not any("conftest" in str(t) for t in target):
-            env.Depends(ninja_file, target)
-        return target, source
-
-    # The "Alias Builder" isn't in the BUILDERS map so we have to
-    # modify it directly.
-    SCons.Environment.AliasBuilder.emitter = ninja_file_depends_on_all
-
-    for _, builder in env["BUILDERS"].items():
-        try:
-            emitter = builder.emitter
-            if emitter is not None:
-                builder.emitter = SCons.Builder.ListEmitter(
-                    [emitter, ninja_file_depends_on_all]
-                )
-            else:
-                builder.emitter = ninja_file_depends_on_all
-        # Users can inject whatever they want into the BUILDERS
-        # dictionary so if the thing doesn't have an emitter we'll
-        # just ignore it.
-        except AttributeError:
-            pass
-
-    # We will subvert the normal Command to make sure all targets generated
-    # from commands will be linked to the ninja file
-    SconsCommand = SCons.Environment.Environment.Command
-
-    def NinjaCommand(self, target, source, action, **kw):
-        targets = SconsCommand(env, target, source, action, **kw)
-        env.Depends(ninja_file, targets)
+        NINJA_STATE.ninja_bin_path = env.get('NINJA_BIN')
+        if not NINJA_STATE.ninja_bin_path:
+            # default to using ninja installed with python module
+            ninja_bin = 'ninja.exe' if env["PLATFORM"] == "win32" else 'ninja'
+            NINJA_STATE.ninja_bin_path = os.path.abspath(os.path.join(
+                ninja_syntax.__file__, 
+                os.pardir, 
+                'data', 
+                'bin', 
+                ninja_bin))
+            if not os.path.exists(NINJA_STATE.ninja_bin_path):
+                # couldn't find it, just give the bin name and hope
+                # its in the path later
+                NINJA_STATE.ninja_bin_path = ninja_bin
+
+    # TODO: this is hacking into scons, preferable if there were a less intrusive way
+    # We will subvert the normal builder execute to make sure all the ninja file is dependent
+    # on all targets generated from any builders 
+    SCons_Builder_BuilderBase__execute = SCons.Builder.BuilderBase._execute
```

Comment:
Why do it this way rather than just writing a scanner?

Reply:
I am looking into the scanner method you suggested currently. I already had this implemented in my local and had been testing with it, so pushed it up before investigating the scanner route a bit more, but there are few draw backs with going through the env's list of builders.

One problem with just using the list of builders in the current environment, is it only captures the builders at the time the ninja tool was generated, so in a given SCons build, if you add builders later, they don't get emitters/scanners updated to make everything depend on the ninja file. 

Secondly, there are other builders that will never be in the env's builder list, such as the Alias builder or the Command builder, so alternate solutions need to be done for those. There may be other builders too that I am not aware of that are never added to the builder list.

This solution, every builder will use this execute function, and this happens after the scons scripts have been read and executed so if you add builders after the ninja tool is initialized.

Ideally the instead of forcing this change directly into the SCons internal class from the tool, SCons has some new function that allows some callback to be setup, or lets you set all emitters/scanners for all builders.